primitives/src/settlement.rs

```diff
@@ -528,57 +529,140 @@ impl InstructionInfo {
 
-        let mut leg_subset = Vec::new();
+        let mut sender_subset = Vec::new();
         for (leg_id, leg) in original_set {
             match leg {
-                Leg::Fungible { sender, .. } | Leg::NonFungible { sender, .. } => {
+                Leg::Fungible {
+                    sender, receiver, ..
+                } => {
+                    if portfolio_set.contains(&sender) {
+                        sender_subset.push((leg_id, leg));
+                        sender_asset_count.add_fungible();
+                    } else if portfolio_set.contains(&receiver) {
+                        receiver_asset_count.add_fungible();
+                    }
+                }
+                Leg::NonFungible {
+                    sender,
+                    receiver,
+                    ref nfts,
+                } => {
                     if portfolio_set.contains(&sender) {
-                        leg_subset.push((leg_id, leg));
+                        sender_asset_count.add_non_fungible(&nfts);
+                        sender_subset.push((leg_id, leg));
+                    } else if portfolio_set.contains(&receiver) {
+                        receiver_asset_count.add_non_fungible(&nfts);
                     }
                 }
                 Leg::OffChain { .. } => continue,
             }
         }
-        let subset_asset_count = AssetCount::from_legs(&leg_subset);
 
         FilteredLegs {
-            leg_subset,
+            sender_side_info: SenderSideInfo::new(sender_subset, sender_asset_count),
+            receiver_asset_count,
             unfiltered_asset_count,
-            subset_asset_count,
         }
     }
 
-    /// Returns a slice of `[(LegId, Leg)]` containing all legs in the subset.
-    pub fn leg_subset(&self) -> &[(LegId, Leg)] {
-        &self.leg_subset
+    /// Returns a slice of `[(LegId, Leg)]` containing all legs in the sender subset.
+    pub fn sender_subset(&self) -> &[(LegId, Leg)] {
+        &self.sender_side_info.sender_subset
     }
 
     /// Returns the [`AssetCount`] for the unfiltered set.
     pub fn unfiltered_asset_count(&self) -> &AssetCount {
         &self.unfiltered_asset_count
     }
 
-    /// Returns the [`AssetCount`] for the subset of legs.
-    pub fn subset_asset_count(&self) -> &AssetCount {
-        &self.subset_asset_count
+    /// Returns the [`AssetCount`] for the sender subset of legs.
+    pub fn sender_asset_count(&self) -> &AssetCount {
+        &self.sender_side_info.sender_asset_count
+    }
+
+    /// Returns the [`AssetCount`] for the receiver side of the instruction.
+    pub fn receiver_asset_count(&self) -> &AssetCount {
+        &self.receiver_asset_count
+    }
+}
+
+/// Holds the [`AssetCount`] for both the sender and receiver side and the number of offchain assets.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Clone, Copy, Debug, Decode, Default, Encode, Eq, PartialEq, TypeInfo)]
+pub struct InputCost {
```

Comment:
`InputCost` is a little bit too generic of a name.  Can use use something like `SettlementAssetCost` or `SettlementCost`?